compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -429,14 +438,45 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
     }
 
     fn before_terminator(ecx: &mut InterpCx<'mir, 'tcx, Self>) -> InterpResult<'tcx> {
-        // The step limit has already been hit in a previous call to `before_terminator`.
-        if ecx.machine.steps_remaining == 0 {
-            return Ok(());
-        }
-
-        ecx.machine.steps_remaining -= 1;
-        if ecx.machine.steps_remaining == 0 {
-            throw_exhaust!(StepLimitReached)
+        if let Some(new_steps) = ecx.machine.num_evaluated_terminators.checked_add(1) {
+            ecx.machine.num_evaluated_terminators = new_steps;
+            if new_steps == LINT_TERMINATOR_LIMIT {
```

Comment:
```suggestion
            // By default, we have a *deny* lint kicking in after some time
            // to ensure `loop {}` doesn't just go forever.
            // In case that lint got reduced, in particular for `--cap-lint` situations, we also
            // have a hard warning shown every now and then for really long executions.
            if new_steps == LINT_TERMINATOR_LIMIT {
```